etl_manager/extract_metadata.py

```diff
@@ -0,0 +1,409 @@
+import cx_Oracle
+
+from etl_manager.utils import write_json, read_json
+
+
+def create_database_connection(settings_file):
+    """Connects to an Oracle database, with settings taken from the specified json file
+
+    The json file should specify:
+        - host: usually localhost
+        - password
+        - port: usually 1521
+        - service_name
+        - user
+
+    Returns a cx_Oracle database connection.
+    """
+    db_settings = read_json(settings_file)
+
+    dsn = cx_Oracle.makedsn(
+        host=db_settings["host"],
+        port=db_settings["port"],
+        service_name=db_settings["service_name"],
+    )
+    connection = cx_Oracle.connect(
+        user=db_settings["user"],
+        password=db_settings["password"],
+        dsn=dsn,
+        encoding="UTF-8",
+    )
+    return connection
+
+
+def get_table_names(database: str, connection) -> list:
+    """Gets names of tables in the database
+
+    You can save the output into a file, or pass it directly to create_table_json
+
+    Parameters
+    ----------
+    database (str):
+        The name of the database/schema - for example DELIUS_ANALYTICS_PLATFORM
+
+    Returns
+    -------
+    List of table names as strings
+
+    """
+    cursor = connection.cursor()
+    cursor.execute(f"SELECT table_name FROM all_tables WHERE OWNER = '{database}'")
+    result = cursor.fetchall()
+    names = [r[0] for r in result]
+    return names
+
+
+def create_json_for_database(
+    description: str, name: str, bucket: str, base_folder: str, location="./metadata",
+):
+    """Creates a database.json file suitable for reading with read_database_folder()
+
+    Parameters
+    ----------
+    description (str):
+        Text saying where the data comes from
+
+    name (str):
+        What you want the database to be called in Glue and Athena. Use underscores
+
+    bucket (str):
+        The S3 bucket where the actual data will end up
+
+    base_folder (str):
+        The folder/prefix within the S3 bucket for the data you want
+        Use capitals if the S3 location includes capitals
+        Don't go down to the table level - it should be the level above
+        For example "hmpps/delius/DELIUS_ANALYTICS_PLATFORM"
+
+    location (str):
+        Folder where you'd like to create the json metadata
+
+    Returns
+    -------
+    None, but creates a json file in the specified location.
+    """
+
+    db = {
+        "description": description,
+        "name": name,
+        "bucket": bucket,
+        "base_folder": base_folder,
+    }
+    write_json(db, f"{location}/database.json")
```

Comment:
Should use `os.path.join` here in case someome provides `my_folder/`

Reply:
Done (and ditto on another place I've used write_json). Do we generally stick with os.path here rather than pathlib? Or is it just whatever fits best in the situation?

Reply:
Historically we've just used os.path but not against pathlib